Panic during reconciling

```
2024-07-26T07:18:29Z    INFO    Observed a panic in reconciler: runtime error: index out of range [0] with length 0     {"controller": "tenantcontrolplane", "controllerGroup": "kamaji.clasti
x.io", "controllerKind": "TenantControlPlane", "TenantControlPlane": {"name":"kubernetes-test4","namespace":"tenant-georg25"}, "namespace": "tenant-georg25", "name": "kubernetes-test4", "rec
oncileID": "3f55c522-6fc9-4ce9-903f-f7809dac8a99"}
panic: runtime error: index out of range [0] with length 0 [recovered]
        panic: runtime error: index out of range [0] with length 0

goroutine 594 [running]:
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:111 +0x1e5
panic({0x2240ec0?, 0xc004f70720?})
        /usr/local/go/src/runtime/panic.go:770 +0x132
github.com/clastix/kamaji/internal/resources.(*KubernetesIngressResource).ShouldStatusBeUpdated(0xc0041b7668?, {0x2747a00?, 0xc00039bf10?}, 0x10?)
        /workspace/internal/resources/k8s_ingress_resource.go:42 +0x179
github.com/clastix/kamaji/internal/resources.createOrUpdate({0x2747a00, 0xc00039bf10}, {0x274f8a0, 0xc00353ce10}, 0xc0028aa008)
        /workspace/internal/resources/resource.go:97 +0x8c
github.com/clastix/kamaji/internal/resources.Handle({0x2747a00, 0xc00039bf10}, {0x274f8a0, 0xc00353ce10}, 0xc0028aa008)
        /workspace/internal/resources/resource.go:67 +0xf2
github.com/clastix/kamaji/controllers.(*TenantControlPlaneReconciler).Reconcile(0xc0007ab440, {0x2747958, 0xc004362ff0}, {{{0xc0048a9290?, 0xe?}, {0xc0048a92d0?, 0x10?}}})
        /workspace/controllers/tenantcontrolplane_controller.go:178 +0xd7e
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x274d9b0?, {0x2747958?, 0xc004362ff0?}, {{{0xc0048a9290?, 0xb?}, {0xc0048a92d0?, 0x0?}}})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:114 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc000329ad0, {0x2747990, 0xc000709ef0}, {0x215eb40, 0xc004342900})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:311 +0x3bc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc000329ad0, {0x2747990, 0xc000709ef0})
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:261 +0x1be
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:222 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 134
        /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.18.4/pkg/internal/controller/controller.go:218 +0x486
```

```
apiVersion: kamaji.clastix.io/v1alpha1
kind: TenantControlPlane
metadata:
  annotations:
    kamaji.clastix.io/kubeconfig-secret-key: super-admin.svc
  creationTimestamp: "2024-07-15T11:57:49Z"
  finalizers:
  - finalizer.kamaji.clastix.io
  - finalizer.kamaji.clastix.io/soot
  generation: 2
  name: kubernetes-test4
  namespace: tenant-georg25
  ownerReferences:
  - apiVersion: controlplane.cluster.x-k8s.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: KamajiControlPlane
    name: kubernetes-test4
    uid: 9098ac17-023e-46d5-8738-023da1903527
  resourceVersion: "127659706"
  uid: e60214b8-2cb4-4293-b482-6709f6ded249
spec:
  addons:
    coreDNS: {}
    konnectivity:
      agent:
        image: registry.k8s.io/kas-network-proxy/proxy-agent
        tolerations:
        - key: CriticalAddonsOnly
          operator: Exists
        version: v0.0.32
      server:
        image: registry.k8s.io/kas-network-proxy/proxy-server
        port: 8132
        version: v0.0.32
  controlPlane:
    deployment:
      additionalMetadata: {}
      additionalVolumeMounts: {}
      extraArgs: {}
      podAdditionalMetadata: {}
      registrySettings:
        apiServerImage: kube-apiserver
        controllerManagerImage: kube-controller-manager
        registry: registry.k8s.io
        schedulerImage: kube-scheduler
      replicas: 2
      resources:
        apiServer: {}
        controllerManager: {}
        kine: {}
        scheduler: {}
      serviceAccountName: default
      strategy: {}
    ingress:
      additionalMetadata:
        annotations:
          nginx.ingress.kubernetes.io/ssl-passthrough: "true"
      hostname: kubernetes-test4.georg25.dev2.example.org:443
      ingressClassName: tenant-root
    service:
      additionalMetadata: {}
      serviceType: ClusterIP
  dataStore: tenant-root
  kubernetes:
    admissionControllers:
    - CertificateApproval
    - CertificateSigning
    - CertificateSubjectRestriction
    - DefaultIngressClass
    - DefaultStorageClass
    - DefaultTolerationSeconds
    - LimitRanger
    - MutatingAdmissionWebhook
    - NamespaceLifecycle
    - PersistentVolumeClaimResize
    - Priority
    - ResourceQuota
    - RuntimeClass
    - ServiceAccount
    - StorageObjectInUseProtection
    - TaintNodesByCondition
    - ValidatingAdmissionWebhook
    kubelet:
      cgroupfs: systemd
      preferredAddressTypes:
      - InternalIP
      - ExternalIP
    version: v1.30.1
  networkProfile:
    certSANs:
    - kubernetes-test4.georg25.dev2.example.org:443
    dnsServiceIPs:
    - 10.95.0.10
    podCidr: 10.243.0.0/16
    port: 6443
    serviceCidr: 10.95.0.0/16
status:
  addons:
    coreDNS:
      enabled: true
      lastUpdate: "2024-07-15T12:02:57Z"
    konnectivity:
      agent:
        lastUpdate: "2024-07-15T16:40:03Z"
        name: konnectivity-agent
        namespace: kube-system
      certificate:
        checksum: 0f14b5d71a0bcb9650169224a10725f0
        lastUpdate: "2024-07-15T12:02:42Z"
        secretName: kubernetes-test4-konnectivity-certificate
      clusterrolebinding:
        name: system:konnectivity-server
      configMap:
        checksum: e8b426b27a2452d0137b14cfbbd10a0d
        name: kubernetes-test4-konnectivity-egress-selector-configuration
      enabled: true
      kubeconfig:
        checksum: 1256aa90f921c682992ce4bde83adf11
        lastUpdate: "2024-07-15T12:02:42Z"
        secretName: kubernetes-test4-konnectivity-kubeconfig
      sa:
        name: konnectivity-agent
        namespace: kube-system
      service:
        loadBalancer: {}
        name: kubernetes-test4
        namespace: tenant-georg25
        port: 8132
    kubeProxy:
      enabled: false
  certificates:
    apiServer:
      checksum: 4244b9b05c7e15c435339a2ae7ceaf75
      lastUpdate: "2024-07-15T12:02:34Z"
      secretName: kubernetes-test4-api-server-certificate
    apiServerKubeletClient:
      checksum: 704cc43f151ee92b578df477d9398f45
      lastUpdate: "2024-07-15T12:02:35Z"
      secretName: kubernetes-test4-api-server-kubelet-client-certificate
    ca:
      checksum: 48052e43d3b298ae71a42e89610a310b
      lastUpdate: "2024-07-15T12:02:30Z"
      secretName: kubernetes-test4-ca
    frontProxyCA:
      checksum: 855a25d0c2869ff46c9baf5b258d0595
      lastUpdate: "2024-07-15T12:02:31Z"
      secretName: kubernetes-test4-front-proxy-ca-certificate
    frontProxyClient:
      checksum: 11ce6f01aba6dc3564d3be05a10d24fd
      lastUpdate: "2024-07-15T12:02:36Z"
      secretName: kubernetes-test4-front-proxy-client-certificate
    sa:
      checksum: 51283dae15d5fef13ed7a00684f37c48
      lastUpdate: "2024-07-15T12:02:32Z"
      secretName: kubernetes-test4-sa-certificate
  controlPlaneEndpoint: 10.96.251.159:6443
  kubeadmPhase:
    bootstrapToken:
      checksum: 3d01adcee9bd20cecfceb4b308fb64d2
      lastUpdate: "2024-07-15T12:02:55Z"
  kubeadmconfig:
    checksum: 5b9eb4f135194f977928580f020a590e
    configmapName: kubernetes-test4-kubeadmconfig
    lastUpdate: "2024-07-15T12:02:28Z"
  kubeconfig:
    admin:
      checksum: 555c2eefca6e29d5721d082a750d29d3
      lastUpdate: "2024-07-15T12:02:38Z"
      secretName: kubernetes-test4-admin-kubeconfig
    controllerManager:
      checksum: 8ea13fbf29058c73e0ed49d2e1b96e12
      lastUpdate: "2024-07-15T12:02:39Z"
      secretName: kubernetes-test4-controller-manager-kubeconfig
    scheduler:
      checksum: 8ea13fbf29058c73e0ed49d2e1b96e12
      lastUpdate: "2024-07-15T12:02:40Z"
      secretName: kubernetes-test4-scheduler-kubeconfig
  kubernetesResources:
    deployment:
      availableReplicas: 2
      conditions:
      - lastTransitionTime: "2024-07-15T12:02:42Z"
        lastUpdateTime: "2024-07-15T12:02:53Z"
        message: ReplicaSet "kubernetes-test4-74c5c8d668" has successfully progressed.
        reason: NewReplicaSetAvailable
        status: "True"
        type: Progressing
      - lastTransitionTime: "2024-07-23T15:50:12Z"
        lastUpdateTime: "2024-07-23T15:50:12Z"
        message: Deployment has minimum availability.
        reason: MinimumReplicasAvailable
        status: "True"
        type: Available
      lastUpdate: "2024-07-23T15:50:18Z"
      name: kubernetes-test4
      namespace: tenant-georg25
      observedGeneration: 3
      readyReplicas: 2
      replicas: 2
      selector: kamaji.clastix.io/name=kubernetes-test4
      updatedReplicas: 2
    ingress:
      loadBalancer:
        ingress:
        - ip: 10.96.207.54
      name: kubernetes-test4
      namespace: tenant-georg25
    service:
      loadBalancer: {}
      name: kubernetes-test4
      namespace: tenant-georg25
      port: 6443
    version:
      status: Ready
      version: v1.30.1
  storage:
    certificate:
      checksum: 1179c0733963abcd0c4a7206af397fe8
      lastUpdate: "2024-07-15T12:02:41Z"
      secretName: kubernetes-test4-datastore-certificate
    config:
      checksum: 9d4994bb2207defef228f28716a2f82a
      secretName: kubernetes-test4-datastore-config
    dataStoreName: tenant-root
    driver: etcd
    setup:
      checksum: 9d4994bb2207defef228f28716a2f82a
      lastUpdate: "2024-07-15T12:02:40Z"
      schema: tenant_georg25_kubernetes_test4
      user: tenant_georg25_kubernetes_test4
```

--- a/internal/resources/k8s_ingress_resource.go
+++ b/internal/resources/k8s_ingress_resource.go
@@ -30,20 +30,49 @@ func (r *KubernetesIngressResource) ShouldStatusBeUpdated(_ context.Context, tcp
 	case tcp.Spec.ControlPlane.Ingress == nil && tcp.Status.Kubernetes.Ingress == nil:
 		// No update in case of no ingress in spec, neither in status.
 		return false
-	case tcp.Spec.ControlPlane.Ingress != nil && tcp.Status.Kubernetes.Ingress == nil,
-		// Must be updated when TCP is using an Ingress, and status is not tracking it
-		// or
-		// Must be updated when the status is referring to an Ingress, although spec doesn't.
-		tcp.Spec.ControlPlane.Ingress == nil && tcp.Status.Kubernetes.Ingress != nil:
+	case tcp.Spec.ControlPlane.Ingress != nil && tcp.Status.Kubernetes.Ingress == nil, // TCP is using an Ingress, Status not tracking it
+		tcp.Spec.ControlPlane.Ingress == nil && tcp.Status.Kubernetes.Ingress != nil: // Status tracks an Ingress, Spec doesn't
 		return true
-	case len(r.resource.Status.LoadBalancer.Ingress) > 0 && tcp.Status.Kubernetes.Ingress == nil || tcp.Status.Kubernetes.Ingress.LoadBalancer.Ingress == nil:
-		// Must be updated since missing the Ingress status
-		return true
-	case r.resource.Status.LoadBalancer.Ingress[0].IP != tcp.Status.Kubernetes.Ingress.LoadBalancer.Ingress[0].IP:
-		// Must bne updated, Ingress load balancer IP is slightly different
+	case len(tcp.Status.Kubernetes.Ingress.IngressStatus.LoadBalancer.Ingress) != len(r.resource.Status.LoadBalancer.Ingress):
+		// Mismatch count of tracked LoadBalancer Ingress
 		return true
 	default:
-		return tcp.Status.Kubernetes.Ingress.Name != r.resource.GetName() || tcp.Status.Kubernetes.Ingress.Namespace != r.resource.GetNamespace()
+		statusIngress := tcp.Status.Kubernetes.Ingress.IngressStatus.LoadBalancer.Ingress
+
+		for i, ingress := range r.resource.Status.LoadBalancer.Ingress {
+			if ingress.IP != statusIngress[i].IP {
+				return true
+			}
+
+			if len(ingress.Ports) != len(statusIngress[i].Ports) {
+				return true
+			}
+
+			for p, port := range ingress.Ports {
+				if port.Port != statusIngress[i].Ports[p].Port {
+					return true
+				}
+
+				if port.Protocol != statusIngress[i].Ports[p].Protocol {
+					return true
+				}
+
+				if port.Error == nil && statusIngress[i].Ports[p].Error != nil ||
+					port.Error != nil && statusIngress[i].Ports[p].Error == nil {
+					return true
+				}
+
+				if port.Error == nil && statusIngress[i].Ports[p].Error == nil {
+					continue
+				}
+
+				if *port.Error != *statusIngress[i].Ports[p].Error {
+					return true
+				}
+			}
+		}
+
+		return false
 	}
 }
 

